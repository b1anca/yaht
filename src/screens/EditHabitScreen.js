import React from "react";
import { useParams, useNavigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { unwrapResult } from "@reduxjs/toolkit";
import { useForm } from "react-hook-form";
import { selectHabitById } from "../features/habits/habitSelectors";
import { updateHabit } from "../features/habits/habitActions";
import { H1 } from "../components/Typography";
import Input from "../components/Input";
import Button from "../components/Button";
import Error from "../components/Error";

const EditHabitScreen = () => {
  const { id } = useParams();
  const { error } = useSelector((state) => state.habits);
  const habit = useSelector(selectHabitById(id));
  const dispatch = useDispatch();

  const { register, handleSubmit } = useForm();
  const navigate = useNavigate();

  const submitForm = (data) =>
    dispatch(updateHabit({ id: habit.id, ...data }))
      .then(unwrapResult)
      .then(() => navigate("/habits"));

  return (
    <>
      <H1>Edit habit</H1>
      <form className="max-w-sm" onSubmit={handleSubmit(submitForm)}>
        {error && <Error>{error}</Error>}
        <Input
          type="name"
          required
          {...register("name")}
          label="Name"
          defaultValue={habit.name}
        />
        {/* TODO: color selector */}
        <Button primary type="submit">
          Save habit
        </Button>
      </form>
    </>
  );
};

EditHabitScreen.displayName = "EditHabitScreen";

export default EditHabitScreen;
