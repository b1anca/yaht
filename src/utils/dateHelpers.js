export const getDaysInMonth = (month, year) => {
  const date = new Date(year, month, 1);
  const days = [];
  while (date.getMonth() === month) {
    days.push(new Date(date));
    date.setDate(date.getDate() + 1);
  }
  return days;
};

export const formatDate = (date) => {
  const monthNames = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];
  const day = date.getDate();
  const monthIndex = date.getMonth();
  const year = date.getFullYear();

  return day + " " + monthNames[monthIndex] + " " + year;
};

export const areDatesEqual = (date1, date2) => {
  return (
    date1.getFullYear() === date2.getFullYear() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getDate() === date2.getDate()
  );
};
