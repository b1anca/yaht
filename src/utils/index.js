export const getRangeFromViewportWidth = () => {
  const viewportWidth = window.innerWidth;

  if (viewportWidth <= 390) {
    return 3;
  } else if (viewportWidth <= 500) {
    return 4;
    // sm
  } else if (viewportWidth <= 640) {
    return 6;
    // md
  } else if (viewportWidth <= 768) {
    return 8;
    // lg
  } else if (viewportWidth <= 1024) {
    return 9;
  } else {
    return 10;
  }
};
