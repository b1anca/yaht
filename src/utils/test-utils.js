import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { setupStore } from "../app/store";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }
  return {
    store,
    user: userEvent.setup(),
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
