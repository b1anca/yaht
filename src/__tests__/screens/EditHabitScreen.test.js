import React from "react";
import { waitFor } from "@testing-library/react";
import { renderWithProviders } from "../../utils/test-utils";
import EditHabitScreen from "../../screens/EditHabitScreen";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ id: 1 }),
}));

describe("EditHabitScreen", () => {
  it("should render correctly", async () => {
    const preloadedState = {
      habits: {
        habits: [
          {
            id: 1,
            name: "Reading",
            color: "#75ab74",
            tasks: [{ id: 11, completed_at: "2023-06-02T03:00:00.000Z" }],
            description: "Reading 10 pages",
            overall_progress: "90.0",
          },
        ],
      },
    };
    const { asFragment } = renderWithProviders(<EditHabitScreen />, {
      preloadedState,
    });

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
