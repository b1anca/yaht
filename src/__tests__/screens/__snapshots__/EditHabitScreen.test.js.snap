// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditHabitScreen should render correctly 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="mb-4 font-extrabold tracking-tight text-zinc-600 dark:text-zinc-200 text-2xl text-center"
      level="h4"
    >
      Edit habit
    </h2>
    <form>
      <div
        class="mb-6"
      >
        <label
          class="block text-sm font-semibold leading-6 mb-2 text-zinc-600 dark:text-zinc-200"
          for="name"
        >
          Name
        </label>
        <input
          aria-label="name"
          class="bg-transparent appearance-none text-zinc-600 dark:text-zinc-200 rounded block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-600 ring-1 ring-zinc-400"
          id="name"
          name="name"
          required=""
          type="name"
        />
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block text-sm font-semibold leading-6 mb-2 text-zinc-600 dark:text-zinc-200"
          for="description"
        >
          Description
        </label>
        <input
          aria-label="description"
          class="bg-transparent appearance-none text-zinc-600 dark:text-zinc-200 rounded block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-600 ring-1 ring-zinc-400"
          id="description"
          name="description"
          type="description"
        />
      </div>
      <label
        class="block text-sm font-semibold leading-6 mb-2 text-zinc-600 dark:text-zinc-200"
        for="color"
      >
        Color
      </label>
      <div
        class="react-colorful"
      >
        <div
          class="react-colorful__saturation"
          style="background-color: rgb(0, 0, 0);"
        >
          <div
            aria-label="Color"
            aria-valuetext="Saturation 32%, Brightness 67%"
            class="react-colorful__interactive"
            role="slider"
            tabindex="0"
          >
            <div
              class="react-colorful__pointer react-colorful__saturation-pointer"
              style="top: 32.99999999999999%; left: 32%;"
            >
              <div
                class="react-colorful__pointer-fill"
                style="background-color: rgb(115, 115, 115);"
              />
            </div>
          </div>
        </div>
        <div
          class="react-colorful__hue react-colorful__last-control"
        >
          <div
            aria-label="Hue"
            aria-valuemax="360"
            aria-valuemin="0"
            aria-valuenow="119"
            class="react-colorful__interactive"
            role="slider"
            tabindex="0"
          >
            <div
              class="react-colorful__pointer react-colorful__hue-pointer"
              style="top: 50%; left: 33.05555555555556%;"
            >
              <div
                class="react-colorful__pointer-fill"
                style="background-color: rgb(0, 0, 0);"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="mt-6 flex flex-col items-center"
      >
        <button
          class="opacity-90 hover:opacity-100 text-sm button py-2 px-6 rounded-full inline-flex justify-center text-sm font-semibold py-2.5 px-4 shadow text-zinc-100 bg-sky-600 hover:bg-sky-700 max-w-md w-full mt-6"
          type="submit"
        >
          Save habit
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
